vowpalwabbit/conditional_contextual_bandit.cc

```diff
@@ -0,0 +1,907 @@
+        label_str += ":";
+        label_str += std::to_string(outcome->cost);
+      }
+
+      delim = ",";
+
+      // Stop after 2...
+      if (counter > 1 && slots.size() > 2)
+      {
+        label_str += delim;
+        label_str += "...";
+        break;
+      }
+    }
+    std::ostringstream label_buf;
+    label_buf << std::setw(all.sd->col_current_label) << std::right << std::setfill(' ') << label_str;
+
+    std::string pred_str = "";
+    delim = "";
+    counter = 0;
+    for (auto slot : decision_scores)
+    {
+      counter++;
+      pred_str += delim;
+      pred_str += std::to_string(slot[0].action);
+      delim = ",";
+
+      // Stop after 3...
+      if (counter > 2)
+      {
+        pred_str += delim;
+        pred_str += "...";
+        break;
+      }
+    }
+    std::ostringstream pred_buf;
+    pred_buf << std::setw(all.sd->col_current_predict) << std::right << std::setfill(' ') << pred_str;
+
+    all.sd->print_update(all.holdout_set_off, all.current_pass, label_buf.str(), pred_buf.str(), num_features,
+        all.progress_add, all.progress_arg);
+  }
+}
+
+void output_example(vw& all, ccb& /*c*/, multi_ex& ec_seq)
+{
+  if (ec_seq.size() <= 0)
+    return;
+
+  std::vector<example*> slots;
+  size_t num_features = 0;
+  float loss = 0.;
+
+  // Should this be done for shared, action and slot?
+  for (auto ec : ec_seq)
+  {
+    num_features += ec->num_features;
+
+    if (ec->l.conditional_contextual_bandit.type == CCB::example_type::slot)
+    {
+      slots.push_back(ec);
+    }
+  }
+
+  // What does it mean for not all of the slots to be labeled? Does it become a non-labeled example at that point?
+  // Is it hold out?
+  bool labeled_example = true;
+  auto preds = ec_seq[0]->pred.decision_scores;
+  for (size_t i = 0; i < slots.size(); i++)
+  {
+    auto outcome = slots[i]->l.conditional_contextual_bandit.outcome;
+    if (outcome != nullptr)
+    {
+      float l = CB_ALGS::get_cost_estimate(
+          outcome->probabilities[TOP_ACTION_INDEX], outcome->cost, preds[i][TOP_ACTION_INDEX].action);
+      loss += l * preds[i][TOP_ACTION_INDEX].score;
+    }
+    else
+    {
+      labeled_example = false;
+    }
+  }
+
+  bool holdout_example = labeled_example;
+  for (size_t i = 0; i < ec_seq.size(); i++) holdout_example &= ec_seq[i]->test_only;
+
+  // TODO what does weight mean here?
+  all.sd->update(holdout_example, labeled_example, loss, ec_seq[SHARED_EX_INDEX]->weight, num_features);
+
+  for (auto sink : all.final_prediction_sink)
+    print_decision_scores(sink, ec_seq[SHARED_EX_INDEX]->pred.decision_scores);
+
+  CCB::print_update(all, slots, preds, num_features);
+}
+
+void finish_multiline_example(vw& all, ccb& data, multi_ex& ec_seq)
+{
+  if (ec_seq.size() > 0)
+  {
+    output_example(all, data, ec_seq);
+    CB_ADF::global_print_newline(all);
+  }
+
+  for (auto a_s : ec_seq[0]->pred.decision_scores)
+  {
+    a_s.delete_v();
```

Comment:
Should not be here.

Reply:
I had quite a bit of trouble with leaks when I was reusing memory, so I decided to use alloc+dealloc. I can take another pass at reducing this